librz/bin/format/dex/dex.c

```diff
@@ -1,239 +1,1855 @@
 		}
 	}
-	free(dex->cal_strings);
-	free(dex->strings);
-	free(dex->classes);
-	free(dex->methods);
-	free(dex->types);
-	free(dex->fields);
-	free(dex->protos);
-	free(dex);
+
+	return proto_id;
+
+dex_proto_id_new_fail:
+	free(proto_id);
+	return NULL;
 }
 
-RzBinDexObj *rz_bin_dex_new_buf(RzBuffer *buf) {
-	rz_return_val_if_fail(buf, NULL);
-	RzBinDexObj *bin = RZ_NEW0(RzBinDexObj);
-	int i;
-	struct dex_header_t *dexhdr;
-	if (!bin) {
-		goto fail;
-	}
-	bin->size = rz_buf_size(buf);
-	bin->b = rz_buf_ref(buf);
-	/* header */
-	if (bin->size < sizeof(struct dex_header_t)) {
-		goto fail;
-	}
-	dexhdr = &bin->header;
-
-	if (bin->size < 112) {
-		goto fail;
-	}
-
-	rz_buf_seek(bin->b, 0, RZ_BUF_SET);
-	rz_buf_read(bin->b, (ut8 *)&dexhdr->magic, 8);
-	dexhdr->checksum = rz_buf_read_le32(bin->b);
-	rz_buf_read(bin->b, (ut8 *)&dexhdr->signature, 20);
-	dexhdr->size = rz_buf_read_le32(bin->b);
-	dexhdr->header_size = rz_buf_read_le32(bin->b);
-	dexhdr->endian = rz_buf_read_le32(bin->b);
-	// TODO: this offsets and size will be used for checking,
-	// so they should be checked. Check overlap, < 0, > bin.size
-	dexhdr->linksection_size = rz_buf_read_le32(bin->b);
-	dexhdr->linksection_offset = rz_buf_read_le32(bin->b);
-	dexhdr->map_offset = rz_buf_read_le32(bin->b);
-	dexhdr->strings_size = rz_buf_read_le32(bin->b);
-	dexhdr->strings_offset = rz_buf_read_le32(bin->b);
-	dexhdr->types_size = rz_buf_read_le32(bin->b);
-	dexhdr->types_offset = rz_buf_read_le32(bin->b);
-	dexhdr->prototypes_size = rz_buf_read_le32(bin->b);
-	dexhdr->prototypes_offset = rz_buf_read_le32(bin->b);
-	dexhdr->fields_size = rz_buf_read_le32(bin->b);
-	dexhdr->fields_offset = rz_buf_read_le32(bin->b);
-	dexhdr->method_size = rz_buf_read_le32(bin->b);
-	dexhdr->method_offset = rz_buf_read_le32(bin->b);
-	dexhdr->class_size = rz_buf_read_le32(bin->b);
-	dexhdr->class_offset = rz_buf_read_le32(bin->b);
-	dexhdr->data_size = rz_buf_read_le32(bin->b);
-	dexhdr->data_offset = rz_buf_read_le32(bin->b);
-
-/* strings */
-#define STRINGS_SIZE ((dexhdr->strings_size + 1) * sizeof(ut32))
-	if (dexhdr->strings_size > bin->size) {
-		goto fail;
-	}
-	bin->strings = RZ_NEWS0(ut32, dexhdr->strings_size + 1);
-	if (!bin->strings) {
-		goto fail;
-	}
-	rz_buf_read_at(bin->b, dexhdr->strings_offset, (ut8 *)bin->strings, dexhdr->strings_size * sizeof(ut32));
-	// TODO: this is unnecessary on Big endian machines
-	for (i = 0; i < dexhdr->strings_size; i++) {
-		ut64 offset = dexhdr->strings_offset + (i * sizeof(ut32));
-		if (offset + 4 > bin->size) {
-			break;
-		}
-		bin->strings[i] = rz_read_le32(&bin->strings[i]);
-	}
-	/* classes */
-	// TODO: not sure about if that is needed
-	size_t classes_size = dexhdr->class_size * DEX_CLASS_SIZE;
-	if (dexhdr->class_offset + classes_size >= bin->size) {
-		if (dexhdr->class_offset < bin->size) {
-			classes_size = bin->size - dexhdr->class_offset;
+#define dex_field_id_free free
```

Comment:
because it gives me a good opportunity to understand what is what. and very helpful for writing code, because i can just change it into a function later if i need to.